Add Google Storage support

erikdubbelboer opened the pull request:
The website mentions S3-compatible storage should work with `Google Storage which exposes the same API as S3`. I'm not sure how old this documentation is and if this is even still possible. I could not get this working so I added native support for Google Storage.

We are currently testing and improving this with our own reporting system (1.2 billion requests per day). I thought I already make a pull request to get some community feedback.

This pull request also excludes the correct dependencies for `com.maxmind.geoip2`.

b-slim commented on extensions/google-extensions/src/main/java/io/druid/firehose/google/GoogleBlob.java:
```diff
@@ -0,0 +1,52 @@
+  public GoogleBlob() {
+  }
+
+  public GoogleBlob(String bucket, String path) {
```
ser deser test will fail without proper annotation here.

b-slim replied:
ok my bad with default constructor will pass, but thought most of the druid code will use something like 
```java
@JsonCreator
  public GoogleBlob(@JsonProperty("bucket") String bucket, @JsonProperty("path") String path)
  {
    this.bucket = bucket;
    this.path = path;
  }
```

b-slim commented on extensions/google-extensions/src/main/java/io/druid/firehose/google/GoogleBlob.java:
```diff
@@ -0,0 +1,52 @@
+
+public class GoogleBlob {
+  @JsonProperty
+  @NotNull
```
what does notnull means ? is this generated by IDE ? also you might wanna have it as final if it is constructed by jackson.

erikdubbelboer replied:
This class is an almost direct copy of [AzureBlob.java](https://github.com/druid-io/druid/blob/master/extensions/azure-extensions/src/main/java/io/druid/firehose/azure/AzureBlob.java). Should I also change these things there? Or should I make a different pull request for that?

b-slim replied:
@erikdubbelboer i am sure there is a lot of thing to clean at the code base, and thought the general policy is to change it opportunistically (when we have to touch a file). So if you want you can send a different PR that cleans Azure module + some improvement...

b-slim commented on extensions/google-extensions/src/main/java/io/druid/firehose/google/GoogleBlob.java:
```diff
@@ -0,0 +1,52 @@
+  @NotNull
+  private String path = null;
+
+  public GoogleBlob() {
```
do you need both constructors ? if so you need to use annotate one with `@JsonCreator`

b-slim commented on extensions/google-extensions/src/main/java/io/druid/firehose/google/StaticGoogleBlobStoreFirehoseFactory.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.List;
+
+public class StaticGoogleBlobStoreFirehoseFactory implements FirehoseFactory<StringInputRowParser> {
+  private static final Logger log = new Logger(StaticGoogleBlobStoreFirehoseFactory.class);
```
if it is static please use the appropriate convention `LOG`

erikdubbelboer replied:
This is copied from the other storage classes. Should I also change it in those other 178 classes?
```bash
$ grep 'private static final Logger log' druid -R | wc -l
178
```

gianm replied:
@b-slim we use `log` in a *LOT* of places (more often than `LOG`!) so I think lowercase is ok. In a lot of druid code it's an exception to the "cap statics" rule.

b-slim replied:
@gianm why having exception, this doesn't make sense in my opinion. I don't see why new code checked in is properly formatted.

b-slim commented on extensions/google-extensions/src/main/java/io/druid/storage/google/GoogleAccountConfig.java:
```diff
@@ -0,0 +1,38 @@
+{
+
+  @JsonProperty
+  @NotNull
```
same as above.

b-slim commented on extensions/google-extensions/src/main/java/io/druid/storage/google/GoogleAccountConfig.java:
```diff
@@ -0,0 +1,38 @@
+  @NotNull
+  private String bucket;
+
+  public String getBucket()
```
need annotation here or it will fail ser/desr

b-slim commented on extensions/google-extensions/src/main/java/io/druid/storage/google/GoogleByteSource.java:
```diff
@@ -0,0 +1,50 @@
+public class GoogleByteSource extends ByteSource
+{
+
+  final private Storage storage;
```
please use the same convention `private final` instead of `final private`

erikdubbelboer replied:
Sure, want me to change the other 36 instances as well, maybe in a separate pull request?
```
$ grep 'final private' druid -R | wc -l
36
```

b-slim commented on extensions/google-extensions/src/main/java/io/druid/storage/google/GoogleDataSegmentKiller.java:
```diff
@@ -0,0 +1,65 @@
+public class GoogleDataSegmentKiller implements DataSegmentKiller
+{
+
+  private static final Logger log = new Logger(GoogleDataSegmentKiller.class);
```
LOG

b-slim commented on extensions/google-extensions/src/main/java/io/druid/storage/google/GoogleDataSegmentKiller.java:
```diff
@@ -0,0 +1,65 @@
+  @Inject
+  public GoogleDataSegmentKiller(final Storage storage)
+  {
+    this.storage = storage;
```
it will be good to check that storage is not null

b-slim commented:
@erikdubbelboer i am wondering, there is not any authentication or credentials to connect/push to a GCS ?

erikdubbelboer commented:
@b-slim No, by default it will get the required auth from the `gcloud` tool which is installed and configured by default when running on a compute or appengine instance, or can be manually configured on other machines.

b-slim commented:
@erikdubbelboer, thanks for the answer, so suppose in case i have (i is the druid process) the wrong credentials how i figure that out ? is there a special exception  or any logging to tell me i am using wrong configs  if i am debugging what's going on.

b-slim commented:
need to add UTs.

fjy commented:
@erikdubbelboer is this still WIP?

can we add this module and the docs to extensions-contrib? this way we can just merge it without having to address all of the CR

erikdubbelboer commented:
@fjy We started using hadoop index tasks, so after using hadoop already only #2646 and #2645 were needed for druid to support google cloud storage through hadoop. I guess using hadoop in between is a bit slower but none of that stuff is time sensitive. The advantage is that druid in combination with hadoop, and hadoop with google cloud storage are much more tested than the code I wrote for this pull request.

Before this I did some testing where instead of #2645 I had hadoop index tasks with "gs://" output schemes create segments which the historical node could directly load using the code in this pull requests. So only the index tasks would require hadoop, everything else could access google cloud storage directly. I guess this is a bit faster and some people might like not having to rely on hadoop for their historical nodes to work.

So I guess one issue is, if you want to merge this, what should happend with "gs://" schemes? Should they generate hadoop segments like with #2645 or should they generate segments that can directly be downloaded using this pull request?

So I'm not sure what I should do with this pull request. I don't know if anyone is really interested in it. Maybe it's a better idea if I just contribute some documentation on how to use hadoop for google cloud storage like we do now?

sandepudi commented:
@erikdubbelboer We'd love to have direct GCS support without the overhead of maintaining a Hadoop cluster. #2645 already helps us quite a bit that we needn't keep scaling HDFS on our Data Proc cluster for historical segments. But this will allow us to even scale down the Hadoop nodes and not worry about affecting queries.

erikdubbelboer commented:
@sandepudi I'm not sure how queries have anything to do with HDFS? When you're using `gs://` and aren't using hadoop for indexing you basically only need one node to "proxy" the segment requests.

It shouldn't be too hard to get this pull request working with the latest changes. But the reason we didn't choose for this route is that it isn't tested as well as the HDFS code. If you want I can bring the pull request up to date and see if there is interest in merging?

cerisier commented:
That would be awesome.

erikdubbelboer commented:
Update: I looked into this code again as we are exploring the option to only start hadoop for our nightly reindex task. One problem I'm running into now is that we already have a lot of segments with a hdfs LoadSpec. We would need to someone convert these to a google load spec. Or we need to somehow have this google storage plugin also handle the hdfs spec segments. I looked into this but I'm not familiar with how these specs are bound to their type and if this can be overwritten for example. Anyone have any suggestions of this is possible? Otherwise is converting all the payloads in mysql and all the descriptor.json files enough to change the load specs?

erikdubbelboer commented:
I just amended this pull request. It's rebased against master and contains some test cases. So some new feedback would be appreciated.

Right now I'm testing converting hdfs segments to google segments using a small python script: https://gist.github.com/erikdubbelboer/55c699b43e4275779b328a8c37d902fd
After that I'll slowly upgrade our cluster to 0.9.1.1 including these changes to test them even more.

erikdubbelboer commented:
I pushed some more bug fixes.

We are now using this code ([on 0.9.1](https://github.com/erikdubbelboer/druid/tree/0.9.1)) in production in the following ways:
- Realtime nodes push segments directly to [GCS](https://cloud.google.com/storage/).
- Historical nodes pull directly from GCS.
- Hadoop index tasks push their results directly to GCS.
- Hadoop index tasks store their logs directly in GCS.
- Overlords pull the index tasks directly from GCS.

All our old segements have been converted using [the python script](https://gist.github.com/erikdubbelboer/55c699b43e4275779b328a8c37d902fd). So now we basically only start our Hadoop cluster once per night to index the previous day.

cerisier commented:
FYI, we were able to make it work perfectly just by adding a
jets3t.properties file with "s3service.s3-endpoint=storage.googleapis.com"
in conf/druid/_common.

*Corentin Kerisit*
VSRE compliant

On Sat, Jul 30, 2016 at 11:48 AM, Erik Dubbelboer <notifications@github.com>
wrote:

> I pushed some more bug fixes.
>
> We are now using this code (on 0.9.1
> <https://github.com/erikdubbelboer/druid/tree/0.9.1>) in production in
> the following ways:
>
>    - Realtime nodes push segments directly to GCS
>    <https://cloud.google.com/storage/>.
>    - Historical nodes pull directly from GCS.
>    - Hadoop index tasks push their results directly to GCS.
>    - Hadoop index tasks store their logs directly in GCS.
>    - Overlords pull the index tasks directly from GCS.
>
> All our old segements have been converted using the python script
> <https://gist.github.com/erikdubbelboer/55c699b43e4275779b328a8c37d902fd>.
> So now we basically only start our Hadoop cluster once per night to index
> the previous day.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/druid-io/druid/pull/2458#issuecomment-236356375>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABEwwi3KAIQF5AgyE2FfDQZv4koxkCe7ks5qax3zgaJpZM4HY9yR>
> .
>

cerisier commented:
FYI, we were able to make it work natively just by adding a `jets3t.properties` file with:

> s3service.s3-endpoint=storage.googleapis.com

in `conf/druid/_common`.

erikdubbelboer commented:
@cerisier That sounds a lot easier. I had no idea this was possible. Do you also use it for all the features I mentioned above?

drcrallen commented:
@cerisier when things get propagated around druid tends to only consider `*.jar` files (see https://github.com/druid-io/druid/issues/1213 ). Did you see this effect?

fjy commented:
@erikdubbelboer is this still WIP?

fjy commented:
Moving this to 0.9.3 to allow for more time for review

erikdubbelboer commented:
We have been using this in production for two weeks now without issues so it seems to be fine. I removed the WIP.

erikdubbelboer commented:
I just pushed a small fix for a bug someone was having and rebased on master.

erikdubbelboer commented:
Fixed a missing case when converting google segments. And again rebased on master.

And I back ported this branch to 0.9.1.1
See: https://github.com/erikdubbelboer/druid/tree/0.9.1.1-google-extensions
For people who want to try these changes with 0.9.1.1 here are my build artifacts:
https://www.atomx.com/druid-0.9.1.1-bin.tar.gz
https://www.atomx.com/druid-google-extensions-0.9.1.1.jar

Test failing has nothing to do with this pull.

nishantmonu51 commented on extensions-contrib/google-extensions/src/main/java/io/druid/storage/google/GoogleDataSegmentPusher.java:
```diff
@@ -0,0 +1,167 @@
+  @Override
+  public String getPathForHadoop()
+  {
+    return null;
```
Looks like this is leading to NPE reported here - 
https://groups.google.com/d/msgid/druid-user/ef36c5ea-8b28-4049-b5f8-fe35b8ea99b8%40googlegroups.com?utm_medium=email&utm_source=footer

erikdubbelboer replied:
@nishantmonu51 The link posted doesn't work?

nishantmonu51 replied:
somehow that message is now deleted. 
New link - https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/druid-user/bd9dXKcJ8tg

erikdubbelboer replied:
I just pushed a change that should fix this.

erikdubbelboer commented:
I rebased on master and fixed all changes needed for #3585
When testing this pull request on 0.9.2 I found that the geoip was again causing conflicts so I modified the exclusions even more to fix it.

fjy commented:
I haven't reviewed this too carefully, but given this is a contrib extension, I am 👍

fjy commented:
@nishantmonu51 can you help with review here?

fjy commented:
@erikdubbelboer minor conflict in docs here

erikdubbelboer commented:
I have rebased on master.

nishantmonu51 commented:
@erikdubbelboer It is still showing conflicts, can you try rebasing again ?

nishantmonu51 commented on docs/content/development/extensions-contrib/google.md:
```diff
@@ -0,0 +1,55 @@
+
+## Deep Storage
+
+[Googel Cloud Storage](https://cloud.google.com/storage/) is another option for deep storage. This requires some additional druid configuration.
```
s/Googel/google/g

erikdubbelboer commented:
@nishantmonu51 fixed and fixed.

nishantmonu51 commented:
@erikdubbelboer Thanks :)

andresgomezfrr commented on docs/content/development/extensions-contrib/google.md:
```diff
@@ -0,0 +1,55 @@
+|Property|Description|Default|Required?|
+|--------|-----------|-------|---------|
+|bucket|Name of the Google Cloud bucket|N/A|yes|
+|path|The path where data is located.|N/A|yes|
```
The correct property is `prefix` instead of `path`.

https://github.com/druid-io/druid/blob/7d36f540e809198206238c93f253ce54e4deed03/extensions-contrib/google-extensions/src/main/java/io/druid/storage/google/GoogleAccountConfig.java#L33

erikdubbelboer replied:
You are right. I'll make a pull request to fix this.
